exonum/src/api/websockets.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! WebSockets
+
+use actix::*;
+use actix_web::ws;
+
+use rand::{self, Rng, ThreadRng};
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::time::Instant;
+
+use api::ServiceApiState;
+use crypto::Hash;
+
+/// WebSocket Message for communication inside websockets part.
+#[derive(Message, Debug)]
+pub struct Message(pub String);
+
+#[derive(Message)]
+#[rtype(usize)]
+pub(crate) struct Subscribe {
+    pub addr: Recipient<Syn, Message>,
+}
+
+#[derive(Message)]
+pub(crate) struct Unsubscribe {
+    pub id: usize,
+}
+
+#[derive(Message)]
+pub(crate) struct Broadcast {
+    pub block_hash: Hash,
+}
+
+pub(crate) struct WsServer {
+    pub(crate) subscribers: HashMap<usize, Recipient<Syn, Message>>,
+    rng: RefCell<ThreadRng>,
+}
+
+impl Default for WsServer {
+    fn default() -> Self {
+        Self {
+            subscribers: HashMap::new(),
+            rng: RefCell::new(rand::thread_rng()),
+        }
+    }
+}
+
+impl Actor for WsServer {
+    type Context = Context<Self>;
+}
+
+impl Handler<Subscribe> for WsServer {
+    type Result = usize;
+
+    fn handle(&mut self, msg: Subscribe, _ctx: &mut Self::Context) -> usize {
+        let id = self.rng.borrow_mut().gen::<usize>();
+        self.subscribers.insert(id, msg.addr);
+
+        id
+    }
+}
+
+impl Handler<Unsubscribe> for WsServer {
+    type Result = ();
+
+    fn handle(&mut self, msg: Unsubscribe, _ctx: &mut Self::Context) {
+        let Unsubscribe { id } = msg;
+        self.subscribers.remove(&id);
+    }
+}
+
+impl Handler<Broadcast> for WsServer {
+    type Result = ();
+
+    fn handle(&mut self, msg: Broadcast, _ctx: &mut Self::Context) {
+        let Broadcast { block_hash } = msg;
+        for addr in self.subscribers.values() {
+            let _ = addr.do_send(Message(format!("Committed new block {:?}", block_hash)));
```

Comment:
Is it okay to ignore possible error here?

Reply:
Generally yes, because error can take such values: `SendError<T>::Full(T)` and `SendError<T>::Closed(T)`. First type can't be applied to our case, and we can't deal with second type, because we can't understand when this error because of closed connection or broken network.